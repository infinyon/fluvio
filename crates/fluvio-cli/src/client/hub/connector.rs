use std::path::PathBuf;
use std::sync::Arc;
use std::fmt::Debug;

use clap::Parser;
use anyhow::Result;

use fluvio_extension_common::Terminal;
use fluvio_hub_util::HUB_API_CONN_LIST;

use crate::{error::CliError};
use crate::common::OutputFormat;

use super::{get_pkg_list, get_hub_access};

/// List available Connectors in the hub
#[derive(Debug, Parser)]
pub enum ConnectorHubSubCmd {
    /// List all available SmartConnectors
    #[clap(name = "list")]
    List(ConnectorHubListOpts),

    /// Download SmartConnector to the local folder
    #[clap(name = "download")]
    Download(ConnectorHubDownloadOpts),
}

impl ConnectorHubSubCmd {
    pub async fn process<O: Terminal + Debug + Send + Sync>(self, out: Arc<O>) -> Result<()> {
        match self {
            ConnectorHubSubCmd::List(opts) => opts.process(out).await,
            ConnectorHubSubCmd::Download(opts) => opts.process(out).await,
        }
    }
}

#[derive(Debug, Parser)]
pub struct ConnectorHubListOpts {
    #[clap(flatten)]
    output: OutputFormat,

    #[clap(long, hide_short_help = true)]
    remote: Option<String>,
}

impl ConnectorHubListOpts {
    pub async fn process<O: Terminal + Debug + Send + Sync>(self, out: Arc<O>) -> Result<()> {
        let pl = get_pkg_list(HUB_API_CONN_LIST, &self.remote).await?;
        output::tableformat(out, pl.packages, self.output.format)?;
        Ok(())
    }
}

#[derive(Debug, Parser)]
pub struct ConnectorHubDownloadOpts {
    /// SmartConnector name: e.g. infinyon/salesforce-sink@v0.0.1
    #[clap(value_name = "name", required = true)]
    package_name: String,

    /// Target local folder or file name
    #[clap(short, long, value_name = "PATH")]
    output: Option<PathBuf>,

    #[clap(long, hide_short_help = true)]
    remote: Option<String>,
}

impl ConnectorHubDownloadOpts {
    pub async fn process<O: Terminal + Debug + Send + Sync>(self, _out: Arc<O>) -> Result<()> {
        let access = get_hub_access(&self.remote)?;

        let package_name = self.package_name;
        let file_name = fluvio_hub_util::cli_pkgname_to_filename(&package_name).map_err(|_| {
            CliError::HubError(format!(
                "invalid package name format {package_name}, is it the form infinyon/json-sql@0.1.0"
            ))
        })?;

        let file_path = if let Some(mut output) = self.output {
            if output.is_dir() {
                output.push(file_name);
            }
            output
        } else {
            PathBuf::from(file_name)
        };
        println!(
            "downloading {package_name} to {}",
            file_path.to_string_lossy()
        );

        let url = fluvio_hub_util::cli_conn_pkgname_to_url(&package_name, &access.remote)
            .map_err(|_| CliError::HubError(format!("invalid pkgname {package_name}")))?;

        let data = fluvio_hub_util::get_package(&url, &access)
            .await
            .map_err(|err| {
                CliError::HubError(format!("downloading {package_name} failed\nServer: {err}"))
            })?;

        std::fs::write(file_path, data).map_err(|err| {
            CliError::Other(format!(
                "unable to write downloaded package to the disk: {err}"
            ))
        })?;
        println!("... downloading complete");
        Ok(())
    }
}

// #[allow(dead_code)]
mod output {

    //!
    //! # Fluvio hub list - output processing
    //!
    //! Format SmartModules response based on output type
    use comfy_table::{Cell, Row};
    use comfy_table::CellAlignment;
    use tracing::debug;
    use serde::Serialize;
    use anyhow::Result;

    use fluvio_extension_common::output::OutputType;
    use fluvio_extension_common::Terminal;
    use fluvio_extension_common::output::TableOutputHandler;
    use fluvio_extension_common::t_println;
    use fluvio_hub_util::PackageMeta;

    #[derive(Serialize)]
    struct ListConnectors(Vec<PackageMeta>);

    // -----------------------------------
    // Format Output
    // -----------------------------------

    /// Format SmartModules based on output type
    pub fn tableformat<O: Terminal>(
        out: std::sync::Arc<O>,
        list_pkgs: Vec<PackageMeta>,
        output_type: OutputType,
    ) -> Result<()> {
        debug!("connectors: {:#?}", list_pkgs);

        if !list_pkgs.is_empty() {
            let connectors = ListConnectors(list_pkgs);
            out.render_list(&connectors, output_type)?;
            Ok(())
        } else {
            t_println!(out, "no connectors");
            Ok(())
        }
    }

    // -----------------------------------
    // Output Handlers
    // -----------------------------------
    impl TableOutputHandler for ListConnectors {
        /// table header implementation
        fn header(&self) -> Row {
            Row::from(["CONNECTOR", "Visibility"])
        }

        /// return errors in string format
        fn errors(&self) -> Vec<String> {
            vec![]
        }

        /// table content implementation
        fn content(&self) -> Vec<Row> {
            self.0
                .iter()
                .map(|e| {
                    Row::from([
                        Cell::new(e.pkg_name()).set_alignment(CellAlignment::Left),
                        Cell::new(&e.visibility).set_alignment(CellAlignment::Left),
                    ])
                })
                .collect()
        }
    }
}
