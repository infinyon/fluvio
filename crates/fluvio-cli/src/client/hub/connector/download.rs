use std::path::PathBuf;
use std::sync::Arc;
use std::fmt::Debug;

use clap::Parser;
use anyhow::Result;

use fluvio_extension_common::Terminal;

use crate::error::CliError;

use super::get_hub_access;

#[derive(Debug, Parser)]
pub struct ConnectorHubDownloadOpts {
    /// SmartConnector name: e.g. infinyon/http-sink@vX.Y.Z
    #[arg(value_name = "name", required = true)]
    package_name: String,

    /// Target local folder or file name
    #[arg(short, long, value_name = "PATH")]
    output: Option<PathBuf>,

    /// Target platform for the package. Optional. By default the host's one is used.
    #[arg(
        long,
        default_value_t = current_platform::CURRENT_PLATFORM.to_string()
    )]
    target: String,

    #[arg(long, hide_short_help = true)]
    remote: Option<String>,
}

impl ConnectorHubDownloadOpts {
    pub async fn process<O: Terminal + Debug + Send + Sync>(self, _out: Arc<O>) -> Result<()> {
        let access = get_hub_access(&self.remote)?;

        let package_name = self.package_name;
        let file_name = fluvio_hub_util::cli_pkgname_to_filename(&package_name).map_err(|_| {
            CliError::HubError(format!(
                "invalid package name format {package_name}, is it the form infinyon/json-sql@0.1.0"
            ))
        })?;

        let file_path = if let Some(mut output) = self.output {
            if output.is_dir() {
                output.push(file_name);
            }
            output
        } else {
            PathBuf::from(file_name)
        };
        println!(
            "downloading {package_name} to {}",
            file_path.to_string_lossy()
        );

        let url =
            fluvio_hub_util::cli_conn_pkgname_to_url(&package_name, &access.remote, &self.target)
                .map_err(|_| CliError::HubError(format!("invalid pkgname {package_name}")))?;

        let data = fluvio_hub_util::get_package(&url, &access)
            .await
            .map_err(|err| {
                CliError::HubError(format!(
                    "downloading {package_name} failed\nHub error: {err}"
                ))
            })?;

        std::fs::write(file_path, data).map_err(|err| {
            CliError::Other(format!(
                "unable to write downloaded package to the disk: {err}"
            ))
        })?;
        println!("... downloading complete");
        Ok(())
    }
}
