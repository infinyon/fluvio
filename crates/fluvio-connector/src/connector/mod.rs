use std::sync::Arc;
use structopt::StructOpt;

mod create;
mod delete;
mod list;

use fluvio::Fluvio;
use crate::error::ConnectorError as ClusterCliError;
use fluvio_extension_common::Terminal;
use fluvio_extension_common::COMMAND_TEMPLATE;

use create::CreateManagedConnectorOpt;
use delete::DeleteManagedConnectorOpt;
use list::ListManagedConnectorsOpt;

#[derive(Debug, StructOpt)]
pub enum ManagedConnectorCmd {
    /// Create a new managed SPU Group
    #[structopt(
        name = "create",
        template = COMMAND_TEMPLATE,
    )]
    Create(CreateManagedConnectorOpt),

    /// Delete a managed SPU Group
    #[structopt(
        name = "delete",
        template = COMMAND_TEMPLATE,
    )]
    Delete(DeleteManagedConnectorOpt),

    /// List all SPU Groups
    #[structopt(
        name = "list",
        template = COMMAND_TEMPLATE,
    )]
    List(ListManagedConnectorsOpt),
}

impl ManagedConnectorCmd {
    pub async fn process<O: Terminal>(
        self,
        out: Arc<O>,
        fluvio: &Fluvio,
    ) -> Result<(), ClusterCliError> {
        match self {
            Self::Create(create) => {
                create.process(fluvio).await?;
            }
            Self::Delete(delete) => {
                delete.process(fluvio).await?;
            }
            Self::List(list) => {
                list.process(out, &fluvio).await?;
            }
        }
        Ok(())
    }
}
