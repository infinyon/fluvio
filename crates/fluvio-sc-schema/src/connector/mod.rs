pub use fluvio_controlplane_metadata::connector::*;

mod convert {

    use crate::{
        AdminSpec, CreatableAdminSpec, DeletableAdminSpec,
        objects::{
            CreateFrom, DeleteRequest, ListRequest, ListResponse, ObjectFrom, ObjectTryFrom,
            WatchRequest, WatchResponse,
        },
    };
    use super::ManagedConnectorSpec;

    impl AdminSpec for ManagedConnectorSpec {}

    impl CreatableAdminSpec for ManagedConnectorSpec {
        const CREATE_TYPE: u8 = 3;
    }

    impl DeletableAdminSpec for ManagedConnectorSpec {
        type DeleteKey = String;
    }

    CreateFrom!(ManagedConnectorSpec, ManagedConnector);
    ObjectFrom!(WatchRequest, ManagedConnector);
    ObjectFrom!(WatchResponse, ManagedConnector);
    ObjectFrom!(ListRequest, ManagedConnector);
    ObjectFrom!(ListResponse, ManagedConnector);
    ObjectFrom!(DeleteRequest, ManagedConnector);

    ObjectTryFrom!(WatchResponse, ManagedConnector);
    ObjectTryFrom!(ListResponse, ManagedConnector);
}
