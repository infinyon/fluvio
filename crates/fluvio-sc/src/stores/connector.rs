pub use fluvio_controlplane_metadata::connector::*;
pub use fluvio_controlplane_metadata::store::k8::K8MetaItem;
