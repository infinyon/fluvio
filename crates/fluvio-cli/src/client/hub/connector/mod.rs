use fluvio_hub_util::cmd::{ConnectorHubDownloadOpts, ConnectorHubListOpts};

use std::sync::Arc;
use std::fmt::Debug;

use clap::Parser;
use anyhow::Result;

use fluvio_extension_common::Terminal;

const CONNECTOR_HUB_DOWNLOAD_REDIRECT_MESSAGE: &str = r#"Connectors running on `InfinyOn Cloud` are automatically downloaded during `fluvio cloud connector create ...`.
Connectors running locally require `cdk` to download and deploy:
1. Install cdk: `fluvio install cdk`
2. Download connector: `cdk hub download ...`
3. Deploy connector: `cdk deploy start ...`"#;

/// List available Connectors in the hub
#[derive(Debug, Parser)]
pub enum ConnectorHubSubCmd {
    #[command(name = "list")]
    List(ConnectorHubListOpts),
    #[command(name = "download")]
    Download(ConnectorHubDownloadOpts),
}

impl ConnectorHubSubCmd {
    pub async fn process<O: Terminal + Debug + Send + Sync>(self, out: Arc<O>) -> Result<()> {
        match self {
            ConnectorHubSubCmd::List(opts) => opts.process(out).await,
            ConnectorHubSubCmd::Download(_) => {
                out.println(CONNECTOR_HUB_DOWNLOAD_REDIRECT_MESSAGE);
                Ok(())
            }
        }
    }
}

#[cfg(test)]
mod test {
    use std::sync::{RwLock, Arc};

    use clap::Parser;
    use fluvio_extension_common::Terminal;
    use fluvio_future::task::run_block_on;

    use crate::client::hub::connector::CONNECTOR_HUB_DOWNLOAD_REDIRECT_MESSAGE;

    use super::ConnectorHubSubCmd;

    #[derive(Default, Debug)]
    struct MockTerminal(Arc<RwLock<String>>);
    impl Terminal for MockTerminal {
        fn print(&self, msg: &str) {
            self.0
                .write()
                .expect("could not print to MockTerminal")
                .push_str(msg);
        }

        fn println(&self, msg: &str) {
            self.0
                .write()
                .expect("could not println to MockTerminal")
                .push_str(&format!("{msg}\n"));
        }
    }

    #[test]
    fn test_calling_fluvio_hub_download_displays_a_help_message() {
        let terminal = Arc::new(MockTerminal::default());
        let cmd =
            ConnectorHubSubCmd::parse_from(["conn", "download", "infinyon/duckdb-sink@0.1.0"]);

        assert!(matches!(cmd, ConnectorHubSubCmd::Download(_)));

        run_block_on(cmd.process(terminal.clone())).expect("command failed to process");

        let x = terminal.0.read().unwrap();
        let x = x.as_str().trim_end();
        assert_eq!(x, CONNECTOR_HUB_DOWNLOAD_REDIRECT_MESSAGE);
    }
}
