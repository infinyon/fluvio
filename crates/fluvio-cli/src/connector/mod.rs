use std::sync::Arc;
use structopt::StructOpt;

use serde::Deserialize;
use std::collections::BTreeMap;
use std::path::PathBuf;
use std::fs::File;
use std::io::Read;

use fluvio::Fluvio;
use fluvio::metadata::connector::{ManagedConnectorSpec, SecretString};
use fluvio_extension_common::Terminal;
use fluvio_extension_common::COMMAND_TEMPLATE;

mod create;
mod update;
mod delete;
mod list;
mod logs;

use create::CreateManagedConnectorOpt;
use update::UpdateManagedConnectorOpt;
use delete::DeleteManagedConnectorOpt;
use list::ListManagedConnectorsOpt;
use logs::LogsManagedConnectorOpt;
use crate::CliError;

#[derive(Debug, StructOpt)]
pub enum ManagedConnectorCmd {
    /// Create a new Managed Connector
    #[structopt(
        name = "create",
        template = COMMAND_TEMPLATE,
    )]
    Create(CreateManagedConnectorOpt),

    /// Update a Managed Connector
    #[structopt(
        name = "update",
        template = COMMAND_TEMPLATE,
    )]
    Update(UpdateManagedConnectorOpt),

    /// Delete a Managed Connector
    #[structopt(
        name = "delete",
        template = COMMAND_TEMPLATE,
    )]
    Delete(DeleteManagedConnectorOpt),

    /// Get the logs for a Managed Connector
    #[structopt(
        name = "logs",
        template = COMMAND_TEMPLATE,
    )]
    Logs(LogsManagedConnectorOpt),

    /// List all Managed Connectors
    #[structopt(
        name = "list",
        template = COMMAND_TEMPLATE,
    )]
    List(ListManagedConnectorsOpt),
}

impl ManagedConnectorCmd {
    pub async fn process<O: Terminal>(self, out: Arc<O>, fluvio: &Fluvio) -> Result<(), CliError> {
        match self {
            Self::Create(create) => {
                create.process(fluvio).await?;
            }
            Self::Update(update) => {
                update.process(fluvio).await?;
            }
            Self::Delete(delete) => {
                delete.process(fluvio).await?;
            }
            Self::Logs(logs) => {
                logs.process().await?;
            }
            Self::List(list) => {
                list.process(out, fluvio).await?;
            }
        }
        Ok(())
    }
}

#[derive(Debug, Deserialize, Clone)]
pub struct ConnectorConfig {
    name: String,
    #[serde(rename = "type")]
    type_: String,
    pub(crate) topic: String,
    pub(crate) version: Option<String>,
    #[serde(default)]
    parameters: BTreeMap<String, String>,
    #[serde(default)]
    secrets: BTreeMap<String, SecretString>,
}

impl ConnectorConfig {
    pub fn from_file<P: Into<PathBuf>>(path: P) -> Result<Self, CliError> {
        let mut file = File::open(path.into())?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        let connector_config: Self = serde_yaml::from_str(&contents)?;
        Ok(connector_config)
    }
}

impl From<ConnectorConfig> for ManagedConnectorSpec {
    fn from(config: ConnectorConfig) -> ManagedConnectorSpec {
        ManagedConnectorSpec {
            name: config.name,
            type_: config.type_,
            topic: config.topic,
            parameters: config.parameters,
            secrets: config.secrets,
            version: config.version,
        }
    }
}

#[test]
fn config_test() {
    let _: ManagedConnectorSpec = ConnectorConfig::from_file("test-data/test-config.yaml")
        .expect("Failed to load test config")
        .into();
}
