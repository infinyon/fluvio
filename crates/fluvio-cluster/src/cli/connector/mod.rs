use std::sync::Arc;
use structopt::StructOpt;

use serde::Deserialize;
use std::collections::BTreeMap;
use std::path::PathBuf;
use std::fs::File;
use std::io::Read;

use fluvio::Fluvio;
use fluvio_controlplane_metadata::connector::ManagedConnectorSpec;
use fluvio_extension_common::Terminal;
use fluvio_extension_common::COMMAND_TEMPLATE;

mod create;
mod delete;
mod list;

use create::CreateManagedConnectorOpt;
use delete::DeleteManagedConnectorOpt;
use list::ListManagedConnectorsOpt;
use crate::cli::ClusterCliError;

#[derive(Debug, StructOpt)]
pub enum ManagedConnectorCmd {
    /// Create a new Managed Connector
    #[structopt(
        name = "create",
        template = COMMAND_TEMPLATE,
    )]
    Create(CreateManagedConnectorOpt),

    /// Delete a Managed Connector
    #[structopt(
        name = "delete",
        template = COMMAND_TEMPLATE,
    )]
    Delete(DeleteManagedConnectorOpt),

    /// List all Managed Connectors
    #[structopt(
        name = "list",
        template = COMMAND_TEMPLATE,
    )]
    List(ListManagedConnectorsOpt),
}

impl ManagedConnectorCmd {
    pub async fn process<O: Terminal>(
        self,
        out: Arc<O>,
        fluvio: &Fluvio,
    ) -> Result<(), ClusterCliError> {
        match self {
            Self::Create(create) => {
                create.process(fluvio).await?;
            }
            Self::Delete(delete) => {
                delete.process(fluvio).await?;
            }
            Self::List(list) => {
                list.process(out, fluvio).await?;
            }
        }
        Ok(())
    }
}

#[derive(Debug, Deserialize, Clone)]
pub struct ConnectorConfig {
    name: String,
    #[serde(rename = "type")]
    type_: String,
    pub(crate) topic: String,
    #[serde(default)]
    pub(crate) create_topic: bool,
    #[serde(default)]
    parameters: BTreeMap<String, String>,
    #[serde(default)]
    secrets: BTreeMap<String, String>,
}

impl ConnectorConfig {
    pub fn from_file<P: Into<PathBuf>>(path: P) -> Result<ConnectorConfig, ClusterCliError> {
        let mut file = File::open(path.into())?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        let connector_config: ConnectorConfig = serde_yaml::from_str(&contents)?;
        Ok(connector_config)
    }
}

impl From<ConnectorConfig> for ManagedConnectorSpec {
    fn from(config: ConnectorConfig) -> ManagedConnectorSpec {
        ManagedConnectorSpec {
            name: config.name,
            type_: config.type_,
            topic: config.topic,
            parameters: config.parameters,
            secrets: config.secrets,
        }
    }
}

#[test]
fn config_test() {
    let _: ManagedConnectorSpec = ConnectorConfig::from_file("test-data/test-config.yaml")
        .expect("Failed to load test config")
        .into();
}
