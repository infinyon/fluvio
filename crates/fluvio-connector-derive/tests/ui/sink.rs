use fluvio_connector_common::connector;

#[connector(sink)]
fn start_fn() {}
