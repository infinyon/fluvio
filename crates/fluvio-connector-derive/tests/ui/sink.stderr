error: Sink Connectors are not implemented yet
 --> tests/ui/sink.rs:3:13
  |
3 | #[connector(sink)]
  |             ^^^^

error[E0601]: `main` function not found in crate `$CRATE`
 --> tests/ui/sink.rs:4:17
  |
4 | fn start_fn() {}
  |                 ^ consider adding a `main` function to `$DIR/tests/ui/sink.rs`
